core/src/main/scala/kafka/tools/GetOffsetShell.scala

```diff
@@ -195,16 +205,24 @@ object GetOffsetShell {
         val upperRange = group(4).map(_.toInt).getOrElse(Int.MaxValue)
         (p: Int) => p >= lowerRange && p < upperRange
     }
-
-    tp => topicFilter.isTopicAllowed(tp.topic, excludeInternalTopics) && partitionFilter(tp.partition)
+    (
+      topic => topicFilter.isTopicAllowed(topic, excludeInternalTopics),
+      tp => topicFilter.isTopicAllowed(tp.topic, excludeInternalTopics) && partitionFilter(tp.partition)
+    )
```

Comment:
The code is getting a bit complicated with all those filter functions. I wonder if we should introduce an abstraction for them now. Something along the lines of the following. The name are not very good but the idea is there. What do you think?

```scala
trait TopicPartitionFilter {
  def isTopicAllowed(topic: String): Boolean
  def isPartitionAllowed(partition: PartitionInfo): Boolean
}

case class UniqueTopicPartitionFilter(
  topicRegex: String,
  partition: Int
) extends TopicPartitionFilter {
 ...
}

case class RangeTopicPartitionFilter(
  topicRegex: String,
  lower: Int,
  upper: Int
) extends TopicPartitionFilter {
  ...
}

case class SetTopicPartitionFilter(
  topic: String,
  partitionIds: Set[Int]
) extends TopicPartitionFilter {
  ...
}

case class CompositePartitionFilter(
  filters: List[TopicPartitionFilter]
) extends TopicPartitionFilter {
  ...
}
```

Reply:
Thank you for this suggestion, I reused the existing `IncludeList` and added a new `PartitionFilter` to simplify this, more suggestions are welcomed.